lib/srv/monitor.go

```diff
@@ -233,3 +231,81 @@ func (ch trackingChannel) Write(buf []byte) (int, error) {
 	ch.t.UpdateClientActivity()
 	return n, err
 }
+
+// TrackingReadConnConfig is a TrackingReadConn configuration.
+type TrackingReadConnConfig struct {
+	//Conn is a client connection.
+	Conn net.Conn
+	// Clock is a clock, realtime or fixed in tests.
+	Clock clockwork.Clock
+	// Context is an external context to cancel the operation.
+	Context context.Context
+	// Cancel is called whenever client context is closed.
+	Cancel context.CancelFunc
+}
+
+// CheckAndSetDefaults checks and sets defaults.
+func (c *TrackingReadConnConfig) CheckAndSetDefaults() error {
+	if c.Conn == nil {
+		return trace.BadParameter("missing parameter Conn")
+	}
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	if c.Context == nil {
+		return trace.BadParameter("missing parameter Context")
+	}
+	if c.Cancel == nil {
+		return trace.BadParameter("missing parameter Cancel")
+	}
+	return nil
+}
+
+// NewTrackingReadConn returns a new tracking read connection.
+func NewTrackingReadConn(cfg TrackingReadConnConfig) (*TrackingReadConn, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &TrackingReadConn{
+		cfg:        cfg,
+		RWMutex:    sync.RWMutex{},
+		Conn:       cfg.Conn,
+		lastActive: time.Time{},
+	}, nil
+}
+
+// TrackingReadConn allows to wrap net.Conn and keeps track of the latest conn read activity.
+type TrackingReadConn struct {
+	cfg TrackingReadConnConfig
+	sync.RWMutex
+	net.Conn
```

Comment:
Can we unexport `sync.RWMutex` and `net.Conn`? You can also just remove `net.Conn` completely probably and just use it from `cfg`.

Reply:
yes, the mtx can be unexpected but `net.Conn` uses struct embedding in order to implement all unnecessary net.Conn methods and unexported net.Conn will require explicitly implementation of net.Conn methods like `RemoteAddr` `SetWriteDeadline` by  the`TrackingReadConn` struct.